### PR TITLE
[HLSL][SPIR-V] Add Vulkan to target triple

### LINKED ISSUE #70051: [SPIR-V] Specify Vulkan target environment in triple
Currently, we assume any physical addressing SPIR-V in the ArchType implies an OpenCL environment, and logical addressing SPIR-V implies a Vulkan environment, however this isn't strictly correct, and we will eventually need a way to specify the specific Vulkan target environment version. We may need to repurpose the SYS/OS part of the triple for this, dropping the Shader Model version when HLSL is targeting Vulkan.

### PR DESCRIPTION
**Pull request by sudonatalie**
Add support for specifying the logical SPIR-V target environment in the triple as Vulkan. When compiling HLSL, this replaces the DirectX Shader Model with a Vulkan environment instead.

Currently, the only supported combinations of SPIR-V version and Vulkan environment are:
- Vulkan 1.2 and SPIR-V 1.5
- Vulkan 1.3 and SPIR-V 1.6

Fixes #70051

### CODE REVIEW

File: llvm/lib/TargetParser/Triple.cpp [hunk 1]
```diff
@@ -1328,6 +1331,31 @@ VersionTuple Triple::getDriverKitVersion() const {
...
 }
 
+VersionTuple Triple::getVulkanVersion() const {
+  if (getArch() != spirv || getOS() != Vulkan)
+    llvm_unreachable("invalid Vulkan SPIR-V triple");
+
+  VersionTuple VulkanVersion = getOSVersion();
+  SubArchType SpirvVersion = getSubArch();
+
+  llvm::DenseMap<VersionTuple, SubArchType> ValidVersionMap = {
```
Comment (Keenuts) [inline llvm/lib/TargetParser/Triple.cpp:1341]: Shall we have a way to differentiate unsupported vulkan version, and unsupported spirv/vulkan mix?
Or have a longer message saying we only support v1.2+spv1.5 and v1.3+spv1.6?)

File: clang/test/Driver/hlsl-lang-targets-spirv.hlsl [hunk 2]
```diff
@@ -3,29 +3,39 @@
...
+// Empty Vulkan environment
 //
 // RUN: not %clang -target spirv %s -S -o /dev/null 2>&1 | FileCheck --check-prefix=CHECK-NO-OS %s
 
-// Invalid shader models
+// Invalid Vulkan environment
 //
-// RUN: not %clang -target spirv--unknown %s -S -o /dev/null 2>&1 | FileCheck --check-prefix=CHECK-BAD-OS %s
+// RUN: not %clang -target spirv--shadermodel %s -S -o /dev/null 2>&1 | FileCheck --check-prefix=CHECK-BAD-OS %s
+// RUN: not %clang -target spirv-unknown-vulkan1.0-compute %s -S -o /dev/null 2>&1 | FileCheck --check-prefix=CHECK-BAD-OS %s
```
Comment (Keenuts) [inline clang/test/Driver/hlsl-lang-targets-spirv.hlsl:19]: I wonder where we should document that we explicitly don't want to support vulkan < 1.2 for now, hence this limit. Maybe here a comment to say "vulkan1.0 is valid, but we chose not to support it?)
Same for the tests which checks invalid spirv/vk mix?

**Review by Keenuts — COMMENTED** [review]
LGTM for the logic, and choices. Just some small nits

**Review by Keenuts — APPROVED** [review]

**Review by michalpaszkowski — APPROVED** [review]

File: clang/test/Driver/hlsl-lang-targets.hlsl [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// REQUIRES: dxil-registered-target
```
Comment (sudonatalie) [inline clang/test/Driver/hlsl-lang-targets.hlsl:1]: This test wasn't being run for me before. I think this change is correct?

**Review by sudonatalie — COMMENTED** [review]
Will wait for a +1 from @llvm-beanz or @bogner since I modified the error messages slightly for DXIL too

**Comment by sudonatalie** [comment]
Rebased due to conflicts. I'm going to go ahead and merge this today to make the next dependent change easier to review, but DXIL folks please still feel free to follow-up if there are any concerns.

(Also, ignoring the code formatter bot since it seems to be an existing intentional formatting choice in the file that doesn't play nicely with clang-format yet)

**Review by llvm-beanz — APPROVED** [review]
Sorry for my delayed review. LGTM.

**Comment by sudonatalie** [comment]
Rebased again and local `check-all` succeeds.

**Pull request merged**